src/wallet/wallet.cpp

```diff
@@ -3024,82 +2982,161 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, int nConfMine, int
         return true;
     }
 
-    if (nTotalLower < nTargetValue)
+    // if the sum of lower coins is lower than target (or equal but with not admissible size)...
+    if ((nTotalLower < nTargetValue) ||
+        (nTotalLower == nTargetValue && selectionTotalBytesLower > availableBytes))
     {
-        if (coinLowestLarger.second.first == NULL)
+        // ...and there is no larger coin, then the problem is unsolvable
+        if (std::get<3>(coinLowestLarger) == 0)
             return false;
-        setCoinsRet.insert(coinLowestLarger.second);
-        nValueRet += coinLowestLarger.first;
+        // ...otherwise return the larger coin
+        vCoinsRet.push_back(std::get<2>(coinLowestLarger));
+        nValueRet = std::get<0>(coinLowestLarger);
+        selectionTotalBytes = std::get<3>(coinLowestLarger);
         return true;
     }
 
-    // Solve subset sum by stochastic approximation
-    sort(vValue.rbegin(), vValue.rend(), CompareValueOnly());
-    vector<char> vfBest;
-    CAmount nBest;
-
-    ApproximateBestSubset(vValue, nTotalLower, nTargetValue, vfBest, nBest, 1000);
-    if (nBest != nTargetValue && nTotalLower >= nTargetValue + CENT)
-        ApproximateBestSubset(vValue, nTotalLower, nTargetValue + CENT, vfBest, nBest, 1000);
+    // this must be kept for preserving ordering between vValue and optimalSelection
+    std::sort(vValue.begin(), vValue.end(), [](tuple<CAmount, CAmount, COutput, size_t> left,
+                                               tuple<CAmount, CAmount, COutput, size_t> right)
+                                               -> bool { return ( std::get<1>(left) > std::get<1>(right)); } );
 
-    // If we have a bigger coin and (either the stochastic approximation didn't find a good solution,
-    //                                   or the next bigger coin is closer), return the bigger coin
-    if (coinLowestLarger.second.first &&
-        ((nBest != nTargetValue && nBest < nTargetValue + CENT) || coinLowestLarger.first <= nBest))
+    std::vector<std::pair<CAmount, size_t>> amountsAndSizes = std::vector<std::pair<CAmount, size_t>>(vValue.size(), std::make_pair(0, 0));
+    for (int i = 0; i < vValue.size(); i++) {
+        amountsAndSizes[i] = std::make_pair(std::get<1>(vValue[i]), std::get<3>(vValue[i]));
+    }
+    
+    const CAmount targetAmount = nTargetValue;
+    const CAmount targetAmountPlusOffsetNoChange = nTargetValue; //no change
+    const CAmount targetAmountPlusOffsetMaxChange = std::get<2>(coinLowestLarger).tx != NULL ?
+                                                    std::min(std::get<1>(coinLowestLarger), nTotalLower) : nTotalLower; //max change
+    const size_t availableTotalSize = availableBytes;
+    std::unique_ptr<CCoinsSelectionAlgorithmBase> bestAlgorithm = nullptr;
+
+    std::unique_ptr<CCoinsSelectionAlgorithmBase> fastNotOptimalAlgorithm;
+    for (int i = 0; i < COINS_SELECTION_INTERMEDIATE_CHANGE_LEVELS + 2; ++i)
```

Comment:
Avoid putting operations in the condition statement. The compiler might be (in most cases) smart enough to understand it evaluates as a constant (such in this case) but this is better
```suggestion
    for (int i = 0, e = COINS_SELECTION_INTERMEDIATE_CHANGE_LEVELS + 2; i < e; ++i)
```

Reply:
I agree with your suggestion.
Modified accordingly.